### DISCUSSION TITLE
Help to set set_output_power ?

### DISCUSSION THREAD

**Discussion by quazardous** [thomluther/anker-solix-api#141] [Q&A]
Hi could someone provide an example to set set_output_power with the API

thx

```json
  "AZV6Y60D34600632": {
    "device_sn": "XXXXX",
    "type": "solarbank",
    "site_id": "XXXXXXXXX",
    "is_admin": false,
    "device_pn": "A17C0",
    "generation": 1,
    "battery_soc": "15",
    "battery_capacity": "1600",
    "battery_energy": "240",
    "charging_power": "0",
    "power_unit": "W",
    "charging_status": "1",
    "charging_status_desc": "bypass",
    "status": "1",
    "status_desc": "online",
    "wireless_type": "1",
    "input_power": "65",
    "output_power": "65",
    "set_output_power": "100",
    "power_cutoff": 10,
    "alias": "BattDav02",
    "data_valid": true,
    "solarbank_count": 2
  },

```

EDIT: this battery is in a power plan with some power schedule, where can I find the schedule rules ?

siteid is set with a the value from the test.py

tried this but get errors
```python
 _out(
        await myapi.apisession.request(
            "post",
            api.API_ENDPOINTS["get_device_parm"],  # pylint: disable=protected-access  # noqa: SLF001
            json={
                "site_id": siteid,
                "param_type": "4",
            },
        )
    )
```

error
```
(10000) Anker Api Error: Failed to request.
Response Text: {"code":10000,"msg":"Failed to request.","trace_id":"53db0cfeb86d09c8d9f5ecf33254b3c0"}
<class 'api.errors.RequestError'>: (10000) Anker Api Error: Failed to request.
Traceback (most recent call last):
  File "/home/david/dev/projects/anker-solix/anker-solix-api/./client.py", line 263, in <module>
    asyncio.run(main())
  File "/usr/lib64/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/david/dev/projects/anker-solix/anker-solix-api/./client.py", line 256, in main
    await testAPI_ENDPOINTS(myapi)
  File "/home/david/dev/projects/anker-solix/anker-solix-api/./client.py", line 178, in testAPI_ENDPOINTS
    await myapi.apisession.request(
  File "/home/david/dev/projects/anker-solix/anker-solix-api/api/session.py", line 399, in request
    errors.raise_error(data)
  File "/home/david/dev/projects/anker-solix/anker-solix-api/api/errors.py", line 109, in raise_error
    raise error(f'({code}) {prefix}: {data.get("msg","Error msg not found")}')
api.errors.RequestError: (10000) Anker Api Error: Failed to request.

```

OK: found a comment -> works only for site owners

I've set up another account for the API... I've to use the main account...

**Comment by thomluther** [comment]
You cannot set the output directly, but you must provide the proper SB schedule structure with the minimum fields.
Easiest way to manipulate actual output is to get the active schedule, parse it for the active day and time slot, and modify the output in the slot.

I use the helper methods set_*_home_load which I build for manipulating the schedule, either for the current slot or for inserting/deleting slots.
Here is a snippet how I test it. When you do not provide a test_schedule object, the actual schedule will be queried from your solarbank and modified accordingly. With a provided test_schedule, only that object will be modified to get the result as it would be sent to the cloud (only for debugging purpose of the method).
```
            new_schedule = await myapi.set_sb2_home_load(
                siteId=siteId,
                deviceSn=deviceSn,
                usage_mode=None,
                preset=None,
                #insert_slot=apitypes.Solarbank2Timeslot(
                #    start_time=datetime.strptime("16:00", "%H:%M"),
                #    end_time=datetime.strptime("19:00", "%H:%M"),
                #    appliance_load=400,
                #    weekdays=None,
                #),
                test_schedule=scheduleSB2,
            )
            if new_schedule:
                common.print_schedule(new_schedule)
            _out(new_schedule)
```

Question: Why are you using the Api directly to modify the output? I think that is too complicated since you have to experience the full Api structure and functionality again. Prepare for many hours of testing and learning....

**Reply by kmpoppe** [reply]
Why would you think that setting the output powerplan wouldn't be something to do with the API?
I was looking for exactly that, I do not use the Anker power measurement device, I have direct access to the 16.7.0 value of the power meter and could set the powerplan at night to output as much or few power as needed according to the development there. 
Apart from the obvious complexity of setting a plan rather than just put a plain value, are there other reasons against that? 
Do you have any experiences with setting output values less than 100 Watts? Do they get ignored and just set to 100 or does the call fail?

**Reply by thomluther** [reply]
> Why would you think that setting the output powerplan wouldn't be something to do with the API?

I was just saying this is complex and much easier if you use the helper methods available in the library.
You can use set_sb2_home_load(siteId=siteId,deviceSn=deviceSn,preset=xxx) and that's it. The helper will automatically find the actual slot and modify it in the plan or creates a plan for you if none exists.

> Do you have any experiences with setting output values less than 100 Watts? Do they get ignored and just set to 100 or does the call fail?

Are you talking about SB1 or SB2? For SB1 you need the set_sb_home_load helper since that is a different plan structure.
The limit are only enforced by the device settings and firmware. You can put anything in the plan and that will also be shown in the App plan, but the device will ignore it and the the plan preset to the active limits.

Furthermore, SB1 has a bug in the Api schedule. Once you apply a plan with a single slot, there won't be anymore any export.
The plan in the App will show 0W although you defined something different. I found no way to fix this condition again via the Api. You must apply a plan change with the App to get export working again.
Therefore when using it for SB1, you need at least 2 slots in the plan before you are going to modify the output setting via the Api. Otherwise it will break sooner or later.

**Comment by quazardous** [comment]
Can I use the `set_device_parm` with a range of schedules  ? but I've seen the helper function too

> Question: Why are you using the Api directly to modify the output? I think that is too complicated since you have to experience the full Api structure and functionality again.

Well long story short: I've 2 e1600 v1 batteries (with 2x2 PV my original system) feeding a e1600 plus with smartmeter (recent upgrade of my system).
So I want to optimize the power schedule to avoid the front batteries to keep too much solar power if the home load is high or to forward too much power if the load is low. The idea is to avoid starving the e1600 plus and using solar bypass when needed or in the other way passing too much unneeded current if front batteries need to charge. It's similar to the smartmeter but iIt will be less reactive but it's some kind of "buffer optimization" every 5 minutes or so.
I was too quick to buy Anker batteries... Never buy 1st generation products lol

EDIT: and I think I will port the needed endpoints on an arduino device so I want to use "pure" endpoints.  Or should I use https://docs.arduino.cc/micropython/ ?

> Prepare for many hours of testing and learning....

that's my favorite lol

I've read somewhere that anker should do something to implement this case

**Reply by thomluther** [reply]
I doubt that Anker will ever provide official Api
For any automation, I'd recommend to use Home Assistant, then you can use my HA integration and setup any automation you want easily in HA.

**Comment by quazardous** [comment]
> I doubt that Anker will ever provide official Api

not an official API but something to make it work togeteher (v1 and v2)

https://www.ankersolix.com/fr/products/a17c13z1?o_1=1&o_2=1&o_2_p=3&o_3=2&o_3_v=3

says 
> Au quatrième trimestre 2024, nous prévoyons que Solarbank 2 E1600 Pro (A17C1) et Solarbank E1600 (A17C0) pourront être utilisés ensemble, de même que Solarbank 2 E1600 Plus (A17C3) et Solarbank E1600 (A17C0).

(In Q4 2024, we expect that Solarbank 2 E1600 Pro (A17C1) and Solarbank E1600 (A17C0) can be used together, as can Solarbank 2 E1600 Plus (A17C3) and Solarbank E1600 (A17C0).)

I don't know what it means

**Reply by thomluther** [reply]
That is an old post, it probably means that SB1 can be configured into same system as SB2 and then will provide some extra settings on the behavior or output power, which is now limited to min. 100/150 W due to various inverter incompatibilities.
When SB1 output is passed to SB2, lower output settings may become possible...we will see.
